### PR TITLE
feat(comm): support source-Partial placements

### PR DESCRIPTION
**Pull request by junjzhang**
## What did you do

Replace the ``NotImplementedError`` from PR #97 with a real implementation: source ``Partial`` placements now flow through the M2M routing + chunk pipeline. Target ``Partial`` is still rejected (cross-PG decomposition isn't well-defined).

**Design (after a few wrong turns; see ``bench/partial_reduce_prototype.py`` for the data)**:

1. **Hand-rolled routing for Partial source** (in ``get_m2m_map``): when source placements contain Partial, we skip the DTensor-encoding/full_tensor routing trace (which collapses Replicate-equivalent peers down to one rank) and compute the m2m_map deterministically from mesh + placements. Each Partial sub-group's (cell, target_owner) ship slots are round-robin assigned to sub-group members so every source rank gets a unique ship task -- no shadow chunks, no redundant transfers, load balanced across peers.

2. **Whole-tensor reduce before chunking** (in ``agent._handle_transfer``'s ``register_tensors`` path): for each source-Partial pair, clone the user's local Partial buffer and run an all-reduce on each Partial sub-group; chunks then point at the reduced buffer. The clone survives in ``BatchState.pair_reduced_sources`` until transfer completes.

   We tried chunk-level reduce first. It's mathematically incorrect for mixed Shard+Partial: peers ship different cells, so reducing their chunk buffers in a single collective mixes different logical positions across peers. The fix would require shadow-chunk participation from non-shipping peers; for Phase 1 we accept the +1x tensor memory cost and leave shadow-chunk optimization to a follow-up.

3. **NCCL sub-group creation** (in ``agent._handle_init_pair``): for each Partial mesh dim, create the corresponding source-side NCCL sub-group via ``get_or_create_process_group`` (cached). Reuses ``local_group`` when the sub-group spans the entire source side (the common 1D-mesh case), avoiding a redundant new_group bootstrap.

**API changes**:
- ``get_m2m_map`` returns a 4-tuple now; the new ``source_partial_reductions: list[(mesh_dim_idx, reduce_op_str)]`` lets callers know which dims need a sub-group reduce. All callers in this PR are updated.
- ``M2MMap`` / ``PairState`` / ``BatchState`` extended with corresponding fields.
- ``init_pair`` docstring + README + ``docs/index.md`` updated to say Partial is supported on the source side.

## New test cases

- ``tests/test_communication_replicate_shard.py``: 4 new Partial cases (1D Partial → Replicate / Shard, 2D Shard+Partial, double Partial), exercising the full chunk pipeline through ``chunk_comm``.
- ``tests/test_partial_chunk_reduce.py`` (new, 21 cases): mesh ndim × reduce_op × dtype × non-contiguous × chunk-size edge sweep, validating Partial-reduce correctness against PyTorch's ``DTensor.redistribute`` ground truth.
- All existing reshard tests still pass (5 + 1 + 11 + 21 + 4 = **42 tests pass**).

## Test results

```
$ pixi run -e dev pytest tests/test_communication_*.py tests/test_partial_chunk_reduce.py --timeout=120
42 passed in 368.40s
```

ruff format + ruff check clean; pre-commit hooks pass.

## Other comments

- ``NotImplementedError`` is still raised in two known sub-cases (Phase 2 followups):
  - Target ``Partial`` (rejected by design)
  - Over-sharded Partial sub-group where ``cells * target_owners < sub_group_size`` (rare in practice)
- Cross-node perf data (``bench/partial_reduce_prototype.py``) shows the design trade-offs that motivated this approach. Chunk-overlap optimization (where reduce/send overlap on independent NVLink/IB paths) is left for a follow-up issue.

## Summary by CodeRabbit

* **New Features**
  * Added support for "Partial" placement strategy in cross-process-group tensor redistribution with automatic source-side reduction.
  * Enhanced benchmarking with support for Partial configurations and improved baseline handling.

* **Bug Fixes**
  * Fixed hardcoded port configuration in distributed tests to use environment variables.

* **Documentation**
  * Updated documentation to clarify how Partial placements are handled across process groups.

* **Tests**
  * Added comprehensive test coverage for Partial-to-Replicate reduction including mixed placements and edge cases.

* **Chores**
  * Updated project gitignore patterns.
  * Added Transformers dependency to development environment.

[Review Change Stack](https://app.coderabbit.ai/change-stack/cmriat/Etha/pull/98?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by junjzhang** [comment]
@codex review

File: src/etha/comm/ir.py [hunk 1]
```diff
@@ -52,6 +65,24 @@ def prepare(self, contiguous: bool = True) -> None:
             buffer = buffer.to(self.transfer_dtype)
```
[the remark this answers is not in the document]
Comment (junjzhang) [reply src/etha/comm/ir.py]: Fixed in 4d1bdf1: `Chunk.prepare` now runs the source-Partial all-reduce in the source dtype before casting to `transfer_dtype`. Added a fp32 source / bf16 target regression test (`test_partial_mixed_dtype`) that fails under the previous reduce-after-cast order.

**Comment by junjzhang** [comment]
@codex review

File: src/etha/tensor_bus/agent.py [hunk 2]
```diff
@@ -304,7 +332,7 @@ def _handle_init_pair(self, msg: InitPair):
                 device="cpu",
             )
             # Second call: second_mesh -> first_mesh
-            map_2, src_slicers_2, tgt_slicers_2 = get_m2m_map(
+            map_2, src_slicers_2, tgt_slicers_2, partial_red_2 = get_m2m_map(
                 source_mesh=second_mesh,
                 source_placements=second_placements,
                 target_mesh=first_mesh,
```
[the remark this answers is not in the document]
Comment (junjzhang) [reply src/etha/tensor_bus/agent.py]: Fixed in 1d2e021. `_handle_init_pair` now skips the direction whose target has Partial (`m2m_send` or `m2m_recv` stays `None`); the other direction works as a unidirectional pair. `register_tensors` builds chunks per available direction; `_handle_transfer` raises a clear error if the user calls the disabled direction. Verified with `tests/test_distributed_model_transfer.py` (non-Partial regression, 2 passed).

**Comment by junjzhang** [comment]
@codex review
